platform-sdk/swirlds-common/src/main/java/com/swirlds/common/utility/throttle/RateLimitedLogger.java

```diff
@@ -0,0 +1,205 @@
+
+    private final Logger logger;
+    private final RateLimiter rateLimiter;
+    private final AutoClosableLock lock;
+
+    /**
+     * Create a new {@link RateLimitedLogger} that will log at most {@code maxFrequency} times per second.
+     *
+     * @param logger       the logger to wrap
+     * @param time         provides wall clock time
+     * @param maxFrequency the maximum frequency at which to log
+     */
+    public RateLimitedLogger(@NonNull final Logger logger, @NonNull final Time time, final double maxFrequency) {
+        this(logger, new RateLimiter(time, maxFrequency));
+    }
+
+    /**
+     * Create a new {@link RateLimitedLogger} that will log at most once per {@code minimumPeriod}.
+     *
+     * @param logger        the logger to wrap
+     * @param time          provides wall clock time
+     * @param minimumPeriod the minimum amount of time that must pass between log messages
+     */
+    public RateLimitedLogger(
+            @NonNull final Logger logger, @NonNull final Time time, @NonNull final Duration minimumPeriod) {
+        this(logger, new RateLimiter(time, minimumPeriod));
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param logger      the logger to wrap
+     * @param rateLimiter the rate limiter to use
+     */
+    private RateLimitedLogger(@NonNull final Logger logger, @NonNull final RateLimiter rateLimiter) {
+        this.logger = Objects.requireNonNull(logger);
+        this.rateLimiter = Objects.requireNonNull(rateLimiter);
+        lock = Locks.createAutoLock();
+    }
+
+    /**
+     * Generate the log message.
+     *
+     * @param baseMessage    the base log message
+     * @param deniedRequests the number of times we have attempted to log this message but have been denied
+     * @return the message to forward to the wrapped logger
+     */
+    @NonNull
+    private static String generateMessage(@NonNull final String baseMessage, final long deniedRequests) {
+        if (deniedRequests > 0) {
+            return baseMessage + "\n(Due to rate limiting, this condition has been triggered "
+                    + deniedRequests
+                    + " times without being reported.)";
+        } else {
+            return baseMessage;
+        }
+    }
+
+    /**
+     * Write a message to the log at debug level. Message may not actually be written if this logger has been used too
+     * recently.
+     *
+     * @param logMarker the marker to use
+     * @param message   the message to write
+     * @param varargs   optional arguments to pass to the log4j logger
+     */
+    public void debug(
+            @NonNull final Marker logMarker, @NonNull final String message, @Nullable final Object... varargs) {
+
+        try (final Locked l = lock.lock()) {
+            final long deniedRequests = rateLimiter.getDeniedRequests();
+            if (rateLimiter.requestAndTrigger()) {
+                logger.debug(logMarker, generateMessage(message, deniedRequests), varargs);
```

Comment:
You can use `log(Level level, Marker marker, String message, Object... params);` By doing so you need to write the method implementation only once.